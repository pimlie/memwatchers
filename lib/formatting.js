import cliui from 'cliui'
import chalk from 'chalk'
import { mean, standardDeviation } from 'simple-statistics'
import { memwatchHeader, heapHeader, heapHeaderPart, diffHeader, diffDetailsHeader } from './constants'

let previousStats = {}
const previousHeapSizes = []
const heapHeaders = []

const prefixes = ['T', 'G', 'M', 'k']
const subprefixes = ['p', 'n', 'u', 'm']

export function humanNumber(num, unit = '', precision = 1, base = 1000) {
  let absNum = Math.abs(parseFloat(num))
  let prefix = ''

  if (absNum >= base) {
    const _prefixes = Array.from(prefixes)

    while (absNum > base) {
      prefix = _prefixes.pop()

      num = num / base
      absNum = Math.abs(num)
    }
  } else if (absNum > 0 && absNum <= 1) {
    const _prefixes = Array.from(subprefixes)

    while (absNum < 1) {
      prefix = _prefixes.pop()

      num = num * base
      absNum = Math.abs(num)
    }
  }

  return `${num.toFixed(precision)}${prefix || unit ? ' ' : ''}${prefix}${unit}`
}

export function humanBytes(num) {
  return humanNumber(num, 'B', 2)
}

export function humanSeconds(time) {
  return humanNumber(time, 's')
}

export function coloredNumber(num, cur, prev) {
  if (cur == prev || prev === undefined) { // eslint-disable-line eqeqeq
    return num
  }

  const color = cur > prev ? 'green' : 'red'
  return chalk[color](num)
}

export function prettyNumber(num, prev) {
  return coloredNumber(humanNumber(num), num, prev)
}

export function prettyBytes(num, prev) {
  return coloredNumber(humanBytes(num), num, prev)
}

export function prettyHeader(cols) {
  let prevPadding = 1

  return cols.map((c, i) => {
    if (!c.text) {
      return c
    }

    const align = c.align || 'left'

    const padCount = c.width - c.text.length
    let padLeft = Math.floor(padCount / 2)
    let padRight = Math.ceil(padCount / 2)

    let hiddenPadLeft = ''
    let hiddenPadRight = ''

    // TODO: the hidden padding as column marker probably doesnt work perfectly
    //  in all cases, but it does work in our use cases
    if (align === 'left' && padRight > 1) {
      padRight--
      hiddenPadRight = chalk.bgBlack.black('|')
      prevPadding = 1
    } else if (align === 'right' && padLeft > 1 && !prevPadding) {
      padLeft--
      hiddenPadLeft = chalk.bgBlack.black('|')
      prevPadding = 0

      if ((padCount % 2) !== 0) {
        padLeft++
        padRight--
      }
    } else {
      prevPadding = 0
    }

    // we want headers to be centered,
    // set left and center with our padding
    c.align = 'left'

    // cliui trims whitespace, use _ and make it disappear ** wow so magic **
    c.text =
      hiddenPadLeft +
      chalk.bgWhite.white('_'.repeat(padLeft)) +
      chalk.bgWhite.black(c.text) +
      chalk.bgWhite.white('_'.repeat(padRight)) +
      hiddenPadRight
    return c
  })
}

export function prettyRow(cols, chalkOption) {
  if (!chalkOption || typeof chalk[chalkOption] === 'undefined') {
    return cols
  }

  if (Array.isArray(cols)) {
    return cols.map(c => (c.text = chalk[chalkOption](c.text)) && c)
  }

  return chalk[chalkOption](cols)
}

export function getHeader(options, stats) {
  const ui = cliui()

  const cols = []

  if (options.averages) {
    if (!heapHeaders.length) {
      heapHeaders.push(...heapHeader)

      if (stats.real_heap_size === undefined) {
        /* istanbul ignore next line */
        heapHeaders.pop()
      }

      options.heapAverages.forEach((heapCount) => {
        const headerParts = [{
          ...heapHeaderPart.mean
        }, {
          ...heapHeaderPart.stddev
        }]

        headerParts[0].text += ` (${heapCount})`
        headerParts[1].text += ` (${heapCount})`

        heapHeaders.push(...headerParts)
      })
    }

    cols.push(...heapHeaders)
  } else {
    for (const key in memwatchHeader) {
      if (stats[key] === undefined) {
        /* istanbul ignore next line */
        continue
      }
      cols.push(memwatchHeader[key] || { text: key })
    }
  }

  // print headers in gray (not prety, too intrusive for here)
  ui.div(...prettyRow(cols, 'gray'))

  return ui.toString()
}

export function getValue(col, key, val) {
  if (col.pretty && col.bytes) {
    return prettyBytes(val, previousStats[key])
  }

  if (col.pretty) {
    return prettyNumber(val, previousStats[key])
  }

  if (col.bytes) {
    return humanBytes(val)
  }

  if (col.number) {
    return humanNumber(val)
  }

  if (col.seconds) {
    // the stats times are in nano seconds, convert to seconds
    return humanSeconds(val / 1e9)
  }

  return val
}

export function getStatLine(stats) {
  const statLine = Object.keys(memwatchHeader).reduce((acc, key) => {
    if (stats[key] === undefined) {
      return acc
    }

    const val = getValue(memwatchHeader[key], key, stats[key])

    return `${acc}${acc ? ' ' : ''}${memwatchHeader[key].text || key}=${val}`
  }, '')

  previousStats = stats
  return statLine
}

export function getStats(options, stats) {
  const ui = cliui()

  const cols = []
  for (const key in memwatchHeader) {
    if (stats[key] === undefined) {
      /* istanbul ignore next line */
      continue
    }

    const col = {
      ...memwatchHeader[key]
    }

    col.text = getValue(col, key, stats[key])

    cols.push(col)
  }
  ui.div(...cols)

  previousStats = stats
  return ui.toString()
}

export function getHeapAverages(options, stats) {
  previousHeapSizes.push(stats.used_heap_size)

  // clean-up not used heap sizes
  previousHeapSizes.splice(0, previousHeapSizes.length - options.maxHeapCount)

  const ui = cliui()

  const cols = []
  cols.push(stats.gcMarkSweepCompactCount)
  cols.push(prettyBytes(stats.used_heap_size, previousStats.used_heap_size))

  if (stats.real_heap_size !== undefined) {
    cols.push(prettyBytes(stats.real_heap_size, previousStats.real_heap_size))
  }

  stats.mean = {}
  options.heapAverages.forEach((heapCount) => {
    if (previousHeapSizes.length >= heapCount) {
      const heapSizes = previousHeapSizes.slice(-1 * heapCount)

      stats.mean[heapCount] = mean(heapSizes)
      cols.push(prettyBytes(stats.mean[heapCount], previousStats.mean[heapCount]))
      cols.push(humanNumber(standardDeviation(heapSizes)))
    }
  })

  ui.div(...cols.map((colVal, index) => {
    const col = {
      ...heapHeaders[index]
    }

    col.text = colVal
    return col
  }))

  previousStats = stats
  return ui.toString()
}

export function getDiff(diff) {
  const ui = cliui()

  // always print headers in gray
  ui.div('')
  ui.div(prettyRow('HEAP DIFF', 'gray'))
  ui.div(...prettyHeader(JSON.parse(diffHeader)))

  for (const type in diff) {
    const headerRow = JSON.parse(diffHeader)

    headerRow[0].text = type

    if (type === 'change') {
      headerRow[1].text = coloredNumber(humanBytes(diff[type].size_bytes), diff.after.size_bytes, diff.before.size_bytes)
      headerRow[2].text = coloredNumber(diff[type].allocated_nodes, diff.after.nodes, diff.before.nodes)
      headerRow[3].text = diff[type].freed_nodes
    } else {
      headerRow[1].text = humanBytes(diff[type].size_bytes)
      headerRow[2].text = diff[type].nodes
      headerRow[3].text = ''

      // dim these rows
      headerRow.forEach(d => (d.text = chalk.dim(d.text)))
    }

    ui.div(...headerRow)
  }

  // make copy also of the objects
  const _diffDetailsHeader = JSON.parse(diffDetailsHeader)

  // calculate max length of what column
  const whatMaxLength = diff.change.details.reduce((acc, d) => Math.max(acc, d.what.length), 0) + 1
  _diffDetailsHeader[0].width = whatMaxLength

  ui.div('')
  ui.div(prettyRow('DETAILS', 'gray'))
  ui.div(...prettyHeader(_diffDetailsHeader))

  const details = diff.change.details
  details.sort((a, b) => {
    if (b.size_bytes !== a.size_bytes) {
      return b.size_bytes - a.size_bytes
    }

    // what should be unique
    return a.what > b.what ? 1 : -1
  })

  for (const detail of details) {
    const detailRow = JSON.parse(diffDetailsHeader)

    detailRow[0].text = detail.what
    detailRow[0].width = whatMaxLength
    detailRow[1].text = humanBytes(detail.size_bytes)
    detailRow[2].text = detail['+']
    detailRow[3].text = detail['-']

    // dim the row if it wasnt growing
    let prettyOption = ''
    if (detail.size_bytes <= 0) {
      prettyOption = 'dim'
    }

    ui.div(...prettyRow(detailRow, prettyOption))
  }

  ui.div('')

  return ui.toString()
}
