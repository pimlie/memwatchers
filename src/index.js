export { getMemwatch } from './memwatch'
export { setOptions, getHeapStats, startStatsInterval } from './utils'
export { startHeapDiff, endHeapDiff, clearHeapDiff } from './heapdiff'
export { default as start } from './start'
