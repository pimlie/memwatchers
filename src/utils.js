import v8 from 'v8'
import consola from 'consola'
import { mean } from 'simple-statistics'
import { defaultOptions } from './constants'

export const logger = consola.withTag('memwatch')

if (logger.level < 4) {
  logger.level = 4
}

let heapMin = Infinity
let heapMax = 0
const ancientHeapSizes = []

export function setOptions(options = {}) {
  /* istanbul ignore next */
  if (options.__mw_checked) {
    return options
  }

  options.__mw_checked = true

  for (const key in defaultOptions) {
    if (options[key] === undefined) {
      options[key] = defaultOptions[key]
    } else if (key === 'heapAverages') {
      // normalize heapAverages option
      if (!Array.isArray(options[key])) {
        options[key] = [options[key]]
      }

      // remove empty values
      options[key] = options[key].filter(c => !!c && c > 1)
      // sort from low to high
      options[key].sort((a, b) => (a - b))
    } else if (key === 'graphSetup' || key === 'graphAddMetric') {
      if (!Array.isArray(options[key])) {
        if (typeof options[key] === 'function') {
          options[key] = [ options[key] ]
        } else {
          logger.warn(`${key} expects an array of functions`)
          options[key] = []
        }
      } else {
        options[key] = options[key].filter(fn => typeof fn === 'function')
      }
    }
  }

  options.maxHeapCount = Math.max(...options.heapAverages)

  if (options.heapDiffOnInterrupt && options.autoHeapDiff) {
    options.autoHeapDiff = false
    logger.warn('autoHeapDiff has been disabled as it conflicts with heapDiffOnInterrupt: true')
  }

  if (options.graph && options.autoHeapDiff) {
    options.autoHeapDiff = false
    logger.warn('autoHeapDiff has been disabled as it conflicts with graph: true')
  }

  if (options.graph && options.useMovingAverage) {
    options.useMovingAverage = 0
    logger.warn('useMovingAverage is ignored with graph: true')
  }

  return options
}

export function getHeapStats(stats) {
  if (!stats) {
    stats = v8.getHeapStatistics()
  }

  // calculate usage_trend similar as memwatch
  ancientHeapSizes.push(stats.used_heap_size)
  ancientHeapSizes.splice(0, ancientHeapSizes.length - 120)

  const ancientHeapSize = mean(ancientHeapSizes)
  const recentHeapSize = mean(ancientHeapSizes.slice(-10))

  stats.usage_trend = Math.round(10 * (recentHeapSize - ancientHeapSize) / ancientHeapSize) / 10
  stats.min_heap_size = heapMin = Math.min(heapMin, stats.total_heap_size)
  stats.max_heap_size = heapMax = Math.max(heapMax, stats.total_heap_size)

  return stats
}

export function startStatsInterval(callback) {
  const hasCallback = typeof callback === 'function'

  let statCounter = 0
  return setInterval(() => {
    const stats = getHeapStats()
    stats.gcMarkSweepCompactCount = ++statCounter

    if (hasCallback) {
      callback(stats)
    }
  }, 1000)
}
