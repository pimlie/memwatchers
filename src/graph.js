import turtleRace from 'turtle-race'
import { getStatLine } from './formatting'

let graph

export function metricColor(perc) {
  perc = parseFloat(perc)

  if (perc > 0.8) {
    return 'red,bold'
  }

  if (perc > 0.6) {
    return 'yellow,bold'
  }

  return ''
}

export function createGraph(options, stats) {
  const graphSetup = {
    seconds: false,
    noAutoStart: false,
    keep: true,
    metrics: {
      heapSize: {
        min: 0,
        aggregator: 'last',
        color: 'green,bold'
      },
      usageTrend: {
        min: 0,
        color: 'yellow,bold'
      },
      maxSize: {
        min: 0,
        aggregator: 'max',
        color: 'green,bold'
      }
    }
  }

  if (options && Array.isArray(options.graphSetup)) {
    options.graphSetup.forEach(fn => fn(graphSetup, stats))
  }

  const turtle = turtleRace(graphSetup)

  graph = {
    turtle,
    heap: turtle.metric(options.groupName, 'heapSize'),
    trend: turtle.metric(options.groupName, 'usageTrend'),
    max: turtle.metric(options.groupName, 'maxSize')
  }

  return graph
}

export function addMetrics(options, stats) {
  /* istanbul ignore next */
  if (!graph) {
    createGraph(options, stats)
  }

  graph.max.push(stats.max_heap_size).color(metricColor(stats.max_heap_size / stats.heap_size_limit))
  graph.heap.push(stats.used_heap_size).color(metricColor(stats.used_heap_size / stats.heap_size_limit))
  graph.trend.push(stats.usage_trend)

  if (options && Array.isArray(options.graphAddMetric)) {
    options.graphAddMetric.forEach(fn => fn(graph.turtle, stats))
  }

  return graph
}

export function setGraphMessage(stats) {
  /* istanbul ignore next */
  if (!graph) {
    return
  }

  graph.turtle.message(`gc(#${process.pid}): ${getStatLine(stats)}`)
}
