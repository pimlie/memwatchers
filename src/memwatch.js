import { logger, startStatsInterval } from './utils'

let memwatch

// provides a memwatch compatible api when the peerDependency is not installed
export function fakeMemwatch() {
  return {
    _faked: true,
    _hasGC: typeof global.gc === 'function',
    heapWarned: false,
    gcWarned: false,
    HeapDiff: function () {
      // TODO: Check if we can also do this with node built-in tools, e.g. chrome devtools protocol?
      this.end = () => {}

      if (!this.heapWarned) {
        logger.warn('Creating heap diffs is only supported when the peerDependency is installed')
        this.heapWarned = true
      }
    },
    gc() {
      if (this._hasGC) {
        /* istanbul ignore next line */
        global.gc()
      } else if (!this.gcWarned) {
        this.gcWarned = true
        logger.warn('You need to run node either with --expose_gc or install the peerDependency to force the garbage collector to run')
      }
    },
    on(event, callback) {
      if (event !== 'stats') {
        return
      }

      this._interval = startStatsInterval(callback)
    },
    off(event) {
      clearInterval(this._interval)
      this._interval = undefined
    }
  }
}

export async function loadMemwatch() {
  try {
    memwatch = await import('@airbnb/node-memwatch').then(m => m.default || m)
  } catch (e) {}

  /* istanbul ignore next */
  if (!memwatch) {
    memwatch = fakeMemwatch()
  }

  return memwatch
}

export async function getMemwatch() {
  if (!memwatch) {
    await loadMemwatch()
  }

  return memwatch
}
