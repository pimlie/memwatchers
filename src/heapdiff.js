import { getMemwatch } from './memwatch'
import { getDiff } from './formatting'
import { logger } from './utils'

let heapDiff

export async function startHeapDiff() {
  logger.info('Dumping first heap (could take a while)')

  const memwatch = await getMemwatch()

  heapDiff = new memwatch.HeapDiff()
  return heapDiff
}

export function endHeapDiff(printStats) {
  if (!heapDiff) {
    return false
  }

  logger.info('Dumping second heap and create diff (could take a while)')
  const diffStats = heapDiff.end()

  if (printStats) {
    logger.log(getDiff(diffStats))
  }

  return diffStats
}

export function clearHeapDiff() {
  if (heapDiff) {
    heapDiff = undefined
  }
}
