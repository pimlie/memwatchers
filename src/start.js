import { mean } from 'simple-statistics'
import { getHeader, getStats, getHeapAverages } from './formatting'
import { addMetrics, setGraphMessage } from './graph'
import { listenInterrupt, doGC } from './interrupt'
import { getMemwatch } from './memwatch'
import { logger, setOptions, getHeapStats, startStatsInterval } from './utils'
import { startHeapDiff, endHeapDiff, clearHeapDiff } from './heapdiff'

export default async function start(options = {}) {
  const memwatch = await getMemwatch()

  if (memwatch._faked) {
    logger.info(`Please install the peer dependency to unlock all of the features`)
  }

  options = setOptions(options)

  logger.success(`${options.appName} is listening for gc stat events`)

  if (options.graph && !options.gcMetrics) {
    // set interval to add metrics to the graph
    // we could also wait for memwatch.stats events to update the graph
    // but as the graph updates independently every second as well
    // it might be better to just align those updates
    startStatsInterval((stats) => {
      addMetrics(options, stats)
    })
  }

  // gcOnInterrupt is default behaviour
  if (options.heapDiffOnInterrupt) {
    listenInterrupt()
  } else if (options.gcOnInterrupt) {
    listenInterrupt(doGC)
  }

  // if neither options are set there is no use to proceed
  // maybe the user only wanted to add a interrupt listener above
  if (!options.verbose && !options.autoHeapDiff) {
    return
  } else if (options.autoHeapDiff && process.env.NODE_ENV === 'production') {
    logger.warn(`Creating heapDiffs is very expensive, only enable this in production if you really have to`)
  }

  const averageGrowth = []
  let previousHeapSize
  let consecutiveGrowth = 0
  let lineCounter

  memwatch.on('stats', (stats) => {
    getHeapStats(stats)

    if (!options.averages && options.useMovingAverage) {
      averageGrowth.push(stats.used_heap_size)
      averageGrowth.slice(averageGrowth.length - options.useMovingAverage)

      stats.real_heap_size = stats.used_heap_size
      stats.used_heap_size = mean(averageGrowth)
    }

    if (previousHeapSize && stats.used_heap_size > previousHeapSize) {
      consecutiveGrowth++
    } else {
      consecutiveGrowth = 0
    }

    previousHeapSize = stats.used_heap_size

    // if enabled we should've intiated a heapDiff at leakGrowthCount - 1
    if (consecutiveGrowth >= options.leakGrowthCount) {
      endHeapDiff(true)

      // reset
      consecutiveGrowth = 0
    } else if (options.autoHeapDiff && consecutiveGrowth >= (options.leakGrowthCount - 1)) {
      // memwatch defines a memleak as 5 consecutive growths
      // we implement the memleak detection here ourselves, so we can
      // start a heapDiff at 5 - 1 growths so we can better check what grows
      startHeapDiff()
    } else {
      clearHeapDiff()
    }

    if (!options.verbose) {
      return
    }

    if (options.graph) {
      if (options.gcMetrics) {
        addMetrics(options, stats)
      }

      setGraphMessage(stats)
      return
    }

    // print a header to console if needed
    if (lineCounter === undefined || lineCounter >= options.headerEveryLines) {
      logger.debug(getHeader(options, stats))
      lineCounter = 1
    } else {
      lineCounter++
    }

    if (options.averages) {
      logger.debug(getHeapAverages(options, stats))
    } else {
      logger.debug(getStats(options, stats))
    }
  })
}
