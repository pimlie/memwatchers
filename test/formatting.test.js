import * as formatting from '../src/formatting'
import gcStats from './utils/gc-stats'
import heapDiff from './utils/heap-diff'

const averageOptions = { averages: true, heapAverages: [2, 3] }

describe('formatting', () => {
  afterEach(() => jest.resetModules())

  test('human bytes', () => {
    const times = [
      [18, '18.00 B'],
      [1800, '1.80 kB'],
      [18000000, '18.00 MB'],
      [1800000000, '1.80 GB']
    ]

    for (const time of times) {
      expect(formatting.humanBytes(time[0])).toBe(time[1])
    }
  })

  test('human seconds', () => {
    const times = [
      [3600, '3.6 ks'], // haha, which human uses kiloseconds?
      [1.1, '1.1 s'],
      [0.011, '11.0 ms'],
      [0.00011, '110.0 us'],
      [0.0000000011, '1.1 ns']
    ]

    for (const time of times) {
      expect(formatting.humanSeconds(time[0])).toBe(time[1])
    }
  })

  test('pretty columns', () => {
    formatting.getStatLine({ key: 2 })
    expect(formatting.getValue({ pretty: true }, 'key', 3)).toMatchSnapshot()
    expect(formatting.getValue({ pretty: true }, 'key', 1)).toMatchSnapshot()
  })

  test('number columns', () => {
    expect(formatting.getValue({ number: true }, 'key', 1)).toMatchSnapshot()
  })

  test('gc stats - getHeader', () => {
    expect(formatting.getHeader({}, gcStats[0])).toMatchSnapshot()
  })

  test('gc stats - getStats', () => {
    expect(formatting.getStats({}, gcStats[0])).toMatchSnapshot()
    expect(formatting.getStats({}, gcStats[1])).toMatchSnapshot()
    expect(formatting.getStats({}, gcStats[2])).toMatchSnapshot()
  })

  test('averages - getHeader', () => {
    expect(formatting.getHeader(averageOptions, gcStats[0])).toMatchSnapshot()
  })

  test('averages - getHeapAverages', () => {
    expect(formatting.getHeapAverages(averageOptions, gcStats[0])).toMatchSnapshot()
    expect(formatting.getHeapAverages(averageOptions, gcStats[1])).toMatchSnapshot()
    expect(formatting.getHeapAverages(averageOptions, gcStats[2])).toMatchSnapshot()
  })

  test('getStatLine', () => {
    expect(formatting.getStatLine(gcStats[0])).toMatchSnapshot()
    expect(formatting.getStatLine(gcStats[1])).toMatchSnapshot()
    expect(formatting.getStatLine(gcStats[2])).toMatchSnapshot()
  })

  test('getDiff', () => {
    expect(formatting.getDiff(heapDiff)).toMatchSnapshot()
  })
})
