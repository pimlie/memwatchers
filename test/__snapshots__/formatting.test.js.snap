// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatting averages - getHeader 1`] = `"[90m# gc[39m       [90mcurrent[39m        [90mreal[39m    [90mmean (2)[39m       [90mÏƒ (2)[39m    [90mmean (3)[39m       [90mÏƒ (3)[39m"`;

exports[`formatting averages - getHeapAverages 1`] = `"4        [31m128.52 MB[39m   [31m128.52 MB[39m"`;

exports[`formatting averages - getHeapAverages 2`] = `"5        [31m127.95 MB[39m   [31m127.38 MB[39m   128.24 MB     283.8 k"`;

exports[`formatting averages - getHeapAverages 3`] = `"6        [32m130.21 MB[39m   [32m134.74 MB[39m   [32m129.08 MB[39m       1.1 M   128.89 MB     961.6 k"`;

exports[`formatting gc stats - getHeader 1`] = `"[90mfull[39m  [90mincr[39m  [90mscavg[39m [90mweak[39m      [90mtime[39m   [90mtrend[39m     [90mcurrent[39m        [90mreal[39m         [90mmin[39m         [90mmax[39m"`;

exports[`formatting gc stats - getStats 1`] = `"4     3     48    2      21.1 us       0   128.52 MB   128.52 MB   184.35 MB   184.35 MB"`;

exports[`formatting gc stats - getStats 2`] = `"5     3     48    2      64.7 ms       0   [31m127.95 MB[39m   [31m127.38 MB[39m   158.66 MB   184.35 MB"`;

exports[`formatting gc stats - getStats 3`] = `"6     3     53    2      46.3 ms       0   [32m130.21 MB[39m   [32m134.74 MB[39m   155.52 MB   184.35 MB"`;

exports[`formatting getDiff 1`] = `
"
[90mHEAP DIFF[39m
          [47m[37m___[39m[49m[47m[30msize[39m[49m[47m[37m___[39m[49m[40m[30m|[39m[49m[47m[37m__[39m[49m[47m[30mnodes[39m[49m[47m[37m__[39m[49m[40m[30m|[39m[49m[47m[37m__[39m[49m[47m[30mfreed[39m[49m[47m[37m__[39m[49m
[2mbefore[22m     [2m161.99 MB[22m   [2m1568606[22m
[2mafter[22m      [2m161.98 MB[22m   [2m1568653[22m
change      [31m-6.76 kB[39m      [32m1383[39m      1336

[90mDETAILS[39m
[47m[37m______[39m[49m[47m[30mwhat[39m[49m[47m[37m______[39m[49m[40m[30m|[39m[49m[47m[37m___[39m[49m[47m[30msize[39m[49m[47m[37m___[39m[49m[40m[30m|[39m[49m[47m[37m____[39m[49m[47m[30m+[39m[49m[47m[37m____[39m[49m[40m[30m|[39m[49m[47m[37m____[39m[49m[47m[30m-[39m[49m[47m[37m____[39m[49m
Code                1.43 kB         2         0
Socket             264.00 B         1         0
WritableState      240.00 B         1         0
ReadableState      216.00 B         1         0
String             216.00 B         5         1
Object             208.00 B         6         2
Timeout            112.00 B         2         1
Closure             88.00 B        12        10
TickObject          56.00 B         1         0
BufferList          48.00 B         1         0
[2mSignal[22m               [2m0.00 B[22m         [2m1[22m         [2m1[22m
[2mTimersList[22m           [2m0.00 B[22m         [2m1[22m         [2m1[22m
[2mNumber[22m             [2m-16.00 B[22m         [2m4[22m         [2m5[22m
[2mTCP[22m                [2m-32.00 B[22m         [2m1[22m         [2m2[22m
[2mShutdownWrap[22m       [2m-64.00 B[22m         [2m0[22m         [2m1[22m
[2mPromise[22m            [2m-96.00 B[22m         [2m0[22m         [2m2[22m
[2msystem / Context[22m  [2m-128.00 B[22m         [2m1[22m         [2m3[22m
[2mstartHeapDiff[22m     [2m-160.00 B[22m         [2m0[22m         [2m1[22m
[2mNative[22m            [2m-384.00 B[22m      [2m1243[22m      [2m1245[22m
[2mArray[22m             [2m-10.15 kB[22m        [2m40[22m        [2m16[22m
"
`;

exports[`formatting getStatLine 1`] = `"[90mfull[39m=4 [90mincr[39m=3 [90mscavg[39m=48 [90mweak[39m=2 [90mtime[39m=21.1 us [90mtrend[39m=0 [90mcurrent[39m=[31m128.52 MB[39m [90mreal[39m=[31m128.52 MB[39m [90mmin[39m=184.35 MB [90mmax[39m=184.35 MB"`;

exports[`formatting getStatLine 2`] = `"[90mfull[39m=5 [90mincr[39m=3 [90mscavg[39m=48 [90mweak[39m=2 [90mtime[39m=64.7 ms [90mtrend[39m=0 [90mcurrent[39m=[31m127.95 MB[39m [90mreal[39m=[31m127.38 MB[39m [90mmin[39m=158.66 MB [90mmax[39m=184.35 MB"`;

exports[`formatting getStatLine 3`] = `"[90mfull[39m=6 [90mincr[39m=3 [90mscavg[39m=53 [90mweak[39m=2 [90mtime[39m=46.3 ms [90mtrend[39m=0 [90mcurrent[39m=[32m130.21 MB[39m [90mreal[39m=[32m134.74 MB[39m [90mmin[39m=155.52 MB [90mmax[39m=184.35 MB"`;

exports[`formatting number columns 1`] = `"1.0"`;

exports[`formatting pretty columns 1`] = `"[32m3.0[39m"`;

exports[`formatting pretty columns 2`] = `"[31m1.0[39m"`;
